src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs

```diff
@@ -73,17 +83,64 @@ internal abstract class AbstractSnippetProvider : ISnippetProvider
+            // since all trivia is removed when converted to a TextChange.
+            var snippetWithTriviaDocument = await GetDocumentWithSnippetAndTriviaAsync(snippetDocument, position, syntaxFacts, cancellationToken).ConfigureAwait(false);
+
+            // Adds annotations to inserted snippet to be formatted, simplified, add imports if needed, etc.
+            var formatAnnotatedSnippetDocument = await AddFormatAnnotationAsync(snippetWithTriviaDocument, position, cancellationToken).ConfigureAwait(false);
+
+            // Goes through and calls upon the formatting engines that the previous step annotated.
             var reformattedDocument = await CleanupDocumentAsync(formatAnnotatedSnippetDocument, cancellationToken).ConfigureAwait(false);
+
             var reformattedRoot = await reformattedDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
             var caretTarget = reformattedRoot.GetAnnotatedNodes(_cursorAnnotation).SingleOrDefault();
+            var mainChangeNode = reformattedRoot.GetAnnotatedNodes(_findSnippetAnnotation).SingleOrDefault();
+
+            // All the TextChanges from the original document. Will include any imports (if necessary) and all snippet associated
+            // changes after having been formatted.
             var changes = await reformattedDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+
+            // Gets a listing of the identifiers that need to be found in the snippet TextChange
+            // and their associated TextSpan so they can later be converted into an LSP snippet format.
+            var placeholders = GetRenameLocationsMap(mainChangeNode, syntaxFacts, cancellationToken);
+
+            // All the changes from the original document to the most updated. Will later be
+            // collpased into one collapsed TextChange.
+            var changesArray = changes.ToImmutableArray();
             return new SnippetChange(
-                textChanges: changes.ToImmutableArray(),
-                cursorPosition: GetTargetCaretPosition(syntaxFacts, caretTarget));
+                textChanges: changesArray,
+                cursorPosition: GetTargetCaretPosition(syntaxFacts, caretTarget),
+                placeholders: placeholders);
+        }
+
+        /// <summary>
+        /// Descends into the inserted snippet to add back trivia on every token.
+        /// </summary>
+        protected static SyntaxNode? GenerateElasticTriviaForSyntax(ISyntaxFacts syntaxFacts, SyntaxNode? node)
+        {
+            if (node is null)
+            {
+                return null;
+            }
+
+            var nodeWithTrivia = node.ReplaceTokens(node.DescendantTokens(descendIntoTrivia: true), AddAnnotations);
+
+            return nodeWithTrivia;
+
+            SyntaxToken AddAnnotations(SyntaxToken oldToken, SyntaxToken newToken)
```

Comment:
```suggestion
            static SyntaxToken AddAnnotations(SyntaxToken oldToken, SyntaxToken newToken)
```

also, consider just inlining as a lambda.